nat-lab/tests/test_direct_connection.py

```diff
@@ -76,90 +91,221 @@
         ConnectionTag.DOCKER_OPEN_INTERNET_CLIENT_1,
         DOCKER_OPEN_INTERNET_CLIENT_1_IP,
     ),
+    (
+        UPNP_PROVIDER,
+        ConnectionTag.DOCKER_UPNP_CLIENT_1,
+        ConnectionTag.DOCKER_UPNP_CLIENT_2,
+        DOCKER_UPNP_CLIENT_2_IP,
+    ),
 ]
 
 
-@pytest.mark.asyncio
-@pytest.mark.timeout(150)
-@pytest.mark.parametrize(
-    "endpoint_providers, client1_type, client2_type, reflexive_ip",
-    UHP_conn_client_types,
-)
-async def test_direct_working_paths(
-    endpoint_providers, client1_type, client2_type, reflexive_ip
-) -> None:
-    async with AsyncExitStack() as exit_stack:
-        api = API()
-        (alpha, beta) = api.default_config_two_nodes()
+@dataclass
+class NodeWithMeshConnection:
+    node: Node
+    client: telio.Client
+    conn: Connection
+    conn_track: ConnectionTracker
+    conn_gw: Optional[Connection]
+
+    def _init_(self, node, client, conn, conn_track, conn_gw=None):
+        self.node = node
+        self.client = client
+        self.conn = conn
+        self.conn_track = conn_track
+        self.conn_gw = conn_gw
+
+
+@asynccontextmanager
+async def new_connections_with_mesh_clients(
+    exit_stack: AsyncExitStack,
+    client1_type: ConnectionTag,
+    ep1: List[str],
```

Comment:
nitpick: Would it be possible to rename ep[123] to something that specifies that this is not the first endpoint, but a list of endpoint providers for first node? During the reading of the code I got confused about this part, and this may be nice readability improvement.

But consider this note as low prio, and if you think, that it makes the code cumbersome, feel free to just mark the note as "resolved" yourself.